File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/paging/NewsDataSource.java
package com.andruid.magic.newsdaily.paging;

import androidx.annotation.NonNull;
import androidx.paging.PageKeyedDataSource;

import com.andruid.magic.newsloader.api.NewsLoader;
import com.andruid.magic.newsloader.model.News;

import java.util.List;

import timber.log.Timber;

public class NewsDataSource extends PageKeyedDataSource<Integer, News> {
    static final int PAGE_SIZE = 10;
    private static final int FIRST_PAGE = 1;
    private String country;
    private NewsLoader newsLoader;

    public NewsDataSource(String country) {
        Timber.tag("dslog").d("datasource created");
        newsLoader = new NewsLoader();
        this.country = country;
    }

    @Override
    public void loadInitial(@NonNull LoadInitialParams<Integer> params, @NonNull final LoadInitialCallback<Integer, News> callback) {
        Timber.tag("pagelog").d("load initial");
        newsLoader.loadHeadlines(country, FIRST_PAGE, PAGE_SIZE, new NewsLoader.NewsLoadedListener() {
            @Override
            public void onSuccess(List<News> newsList, boolean hasMore) {
                callback.onResult(newsList, null, FIRST_PAGE + 1);
            }

            @Override
            public void onFailure(Throwable t) {
                t.printStackTrace();
            }
        });
    }

    @Override
    public void loadBefore(@NonNull final LoadParams<Integer> params, @NonNull final LoadCallback<Integer, News> callback) {
        newsLoader.loadHeadlines(country, params.key, PAGE_SIZE, new NewsLoader.NewsLoadedListener() {
            @Override
            public void onSuccess(List<News> newsList, boolean hasMore) {
                Integer adjacentKey = (params.key > FIRST_PAGE) ? params.key - 1 : null;
                if(newsList!=null)
                    callback.onResult(newsList, adjacentKey);
            }

            @Override
            public void onFailure(Throwable t) {
                t.printStackTrace();
            }
        });
    }

    @Override
    public void loadAfter(@NonNull final LoadParams<Integer> params, @NonNull final LoadCallback<Integer, News> callback) {
        Timber.tag("pagelog").d("load page %d", params.key);
        newsLoader.loadHeadlines(country, params.key, PAGE_SIZE, new NewsLoader.NewsLoadedListener() {
            @Override
            public void onSuccess(List<News> newsList, boolean hasMore) {
                Integer key = hasMore ? params.key + 1 : null;
                callback.onResult(newsList, key);
            }

            @Override
            public void onFailure(Throwable t) {
                t.printStackTrace();
            }
        });
    }
}

File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/paging/NewsDataSource.kt
package com.andruid.magic.newsdaily.paging

import androidx.paging.PageKeyedDataSource
import com.andruid.magic.newsloader.api.NewsRepository.Companion.getInstance
import com.andruid.magic.newsloader.data.Constants
import com.andruid.magic.newsloader.model.NewsOnline
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

class NewsDataSource(
    private val scope: CoroutineScope,
    private val country: String,
    private val category: String
) : PageKeyedDataSource<Int, NewsOnline>() {

    override fun loadInitial(
        params: LoadInitialParams<Int>,
        callback: LoadInitialCallback<Int, NewsOnline>
    ) {
        scope.launch {
            try {
                val response = getInstance().loadHeadlines(
                    country = country,
                    category = category,
                    page = Constants.FIRST_PAGE,
                    pageSize = Constants.PAGE_SIZE
                )
                when {
                    response.isSuccessful -> {
                        val newsList = response.body()?.newsOnlineList
                        val hasMore = response.body()?.hasMore
                        callback.onResult(
                            newsList ?: listOf(), null,
                            if (hasMore!!) Constants.FIRST_PAGE + 1 else null
                        )
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun loadAfter(params: LoadParams<Int>, callback: LoadCallback<Int, NewsOnline>) {
        scope.launch {
            try {
                val response = getInstance().loadHeadlines(
                    country = country,
                    category = category,
                    page = params.key,
                    pageSize = Constants.PAGE_SIZE
                )
                when {
                    response.isSuccessful -> {
                        val newsList = response.body()?.newsOnlineList
                        val hasMore = response.body()?.hasMore
                        val key = if (hasMore!!) params.key + 1 else null
                        callback.onResult(newsList ?: listOf(), key)
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun loadBefore(params: LoadParams<Int>, callback: LoadCallback<Int, NewsOnline>) {
        scope.launch {
            try {
                val response = getInstance().loadHeadlines(
                    country = country,
                    category = category,
                    page = params.key,
                    pageSize = Constants.PAGE_SIZE
                )
                when {
                    response.isSuccessful -> {
                        val newsList = response.body()?.newsOnlineList
                        val adjacentKey = if (params.key > Constants.FIRST_PAGE)
                            params.key - 1
                        else null
                        callback.onResult(newsList ?: listOf(), adjacentKey)
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
}